### PR TITLE
#241 Group call notifications

### PR DESCRIPTION
**Pull request by Turubaev**
This PR add ring for group jitsi calls. Also it includes corresponding lab feature, that enable/disable this functionality.

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/call/service/CallHeadsUpActionReceiver.kt
Comment (SpiritCroc) [inline vector/src/main/java/im/vector/app/features/call/service/CallHeadsUpActionReceiver.kt]: Should this really cancel all notifications?

**Comment by SpiritCroc** [comment]
Thanks for your PR! I'll properly review and test when I have more time, maybe on the weekend.
I see you have added strings to both `strings.xml` and `strings_sc.xml`, you probably want to remove the duplicated ones from `strings.xml`.

File: vector/src/main/java/im/vector/app/features/call/service/CallHeadsUpActionReceiver.kt
Comment (Turubaev) [reply vector/src/main/java/im/vector/app/features/call/service/CallHeadsUpActionReceiver.kt]: Yeap, this was for testing purpose. Fixed, here we should cancel certain notification

**Comment by Turubaev** [comment]
> Thanks for your PR! I'll properly review and test when I have more time, maybe on the weekend. I see you have added strings to both `strings.xml` and `strings_sc.xml`, you probably want to remove the duplicated ones from `strings.xml`.

When I try to remove from "string.xml", I got "Unresolved reference: call_notification_open_app_action" while building the project. May be you can give example, how to correctly fix it

File: library/ui-strings/src/main/res/values/strings_sc.xml
Comment (SpiritCroc) [inline library/ui-strings/src/main/res/values/strings_sc.xml:240]: You added the strings within another string, that's why the build can't find them - need to properly define them where they belong

Comment (Turubaev) [reply library/ui-strings/src/main/res/values/strings_sc.xml:240]: Got it, thanks !

Comment (alexander-potemkin) [reply library/ui-strings/src/main/res/values/strings_sc.xml:240]: @SpiritCroc , just in case if that got slipped away unnoticed - strings [has been adjusted](https://github.com/SchildiChat/SchildiChat-android/pull/250/commits/00668f80d7477de9efa2b3e3bf2de77cc0b329e5) 

Please, let us know if any other obstacles or things to be adjusted / improved!

Comment (SpiritCroc) [reply library/ui-strings/src/main/res/values/strings_sc.xml:240]: I'm aware of that, thanks. The main thing I'm lacking right now is time to properly test and review, I assume earliest in two week. I usually don't do video calls on mobile, but until https://github.com/element-hq/element-android/issues/8885 is fixed it's probably not too meaningful to get this released anyway.

Comment (alexander-potemkin) [reply library/ui-strings/src/main/res/values/strings_sc.xml:240]: Thank you for a quick response and sharing what's on the plate.

We didn't have / tested video calls in mind as well - do you believe that issue [8885](https://github.com/element-hq/element-android/issues/8885) is a blocker for audio calls as well?

**Review by SpiritCroc — CHANGES_REQUESTED** [review]
Would you mind rebasing on top of SchildiChat's `sc` branch? If I merge this as is currently, it pulls in some unwanted upstream commits

File: vector/src/main/java/im/vector/app/features/MainActivity.kt
Comment (SpiritCroc) [inline vector/src/main/java/im/vector/app/features/MainActivity.kt]: Can you change this back to `} else {` ?

File: library/ui-strings/src/main/res/values/strings_sc.xml
Comment (SpiritCroc) [reply library/ui-strings/src/main/res/values/strings_sc.xml:240]: If it uses the same foreground permission I assume it will be the same, I would assume both use similar implementations but I haven't checked

File: vector/src/main/java/im/vector/app/features/MainActivity.kt
Comment (Turubaev) [reply vector/src/main/java/im/vector/app/features/MainActivity.kt]: Sure, done

**Comment by Turubaev** [comment]
> Would you mind rebasing on top of SchildiChat's `sc` branch? If I merge this as is currently, it pulls in some unwanted upstream commits

Rebased, could you check if it is correct now ?

File: vector-config/src/main/res/values/config-settings.xml
Comment (SpiritCroc) [inline vector-config/src/main/res/values/config-settings.xml]: Should be false

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/events/model/Event.kt
Comment (SpiritCroc) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/events/model/Event.kt]: This looks too generic to me, won't that trigger for all kinds of widgets, not just jitsi?

File: changelog.d/7761.feature
Comment (SpiritCroc) [inline changelog.d/7761.feature]: We don't need this file for downstream changes

File: library/ui-strings/src/main/res/values-ru/strings_sc.xml
Comment (SpiritCroc) [inline library/ui-strings/src/main/res/values-ru/strings_sc.xml]: I'm not sure if weblate will be happy for manually added translations. Maybe it works but safer path would probably merging without translation first, then add translations via weblate later. :thinking:

**Review by SpiritCroc — CHANGES_REQUESTED** [review]
Rebase is fine!
I think your logic for detecting jitsi calls is too generic. From checking the code it looks like it triggers for all kind of widgets. From my experiments it also triggers when removing widgets (i.e. the call will ring again when you actually remove it).

In case it's too hard to get the detection fully accurate, you may get a way with adding a description to the settings so users know about the downsides, but showing a call notification for say a timetable widget being added could be somewhat confusing.

**Comment by SpiritCroc** [comment]
This is quite buggy. I now get call notifications even when I send myself messages in a total different room (a DM), for a different room in which I stopped the widget before.

**Comment by alexander-potemkin** [comment]
> This is quite buggy. I now get call notifications even when I send myself messages in a total different room (a DM), for a different room in which I stopped the widget before.

Apologies for the issue and thanks for raising it - somehow it missed our tests - something we shall work at.

Meanwhile - picking this up, hopefully will be fixed in a week or two max.

File: library/ui-strings/src/main/res/values-ru/strings_sc.xml
Comment (Turubaev) [reply library/ui-strings/src/main/res/values-ru/strings_sc.xml]: Removed manually added translation for Russian

File: changelog.d/7761.feature
Comment (Turubaev) [reply changelog.d/7761.feature]: Removed this file

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/events/model/Event.kt
Comment (Turubaev) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/events/model/Event.kt]: Updated detecting function to more accurate one

File: vector-config/src/main/res/values/config-settings.xml
Comment (Turubaev) [reply vector-config/src/main/res/values/config-settings.xml]: Disabled by default

**Comment by Turubaev** [comment]
> This is quite buggy. I now get call notifications even when I send myself messages in a total different room (a DM), for a different room in which I stopped the widget before.

Fixed all cases, tested, now that should be much better. Could you please check updated version?

File: library/external/textdrawable/build.gradle
Comment (SpiritCroc) [inline library/external/textdrawable/build.gradle:3]: How does this change relate to the PR?
(I see the line is duplicated, but fixing upstream code style is not really something we want to do here)

**Review by SpiritCroc — COMMENTED** [review]
Thanks! With these changes, it no longer notified for jitsi calls for me. Looks like I got
`JitsiEventContent(type=jitsi, name=Jitsi)`, i.e. with uppercase name rather then what you're checking. I was able to fix by adding `lowercase()` calls to the check, so it matches both lowercase and uppercase scenarios:
```
content?.toModel<JitsiEventContent>()?.type?.lowercase() == IS_JITSI_CALL && content.toModel<JitsiEventContent>()?.name?.lowercase() == IS_JITSI_CALL
```

Otherwise seems to work fine now. I think I'm fine with merging, but I'll probably want to do some code refactoring right after to reduce possible future merge conflicts (I've learnt some tricks which probably doesn't make sense to expect from new contributors to apply).

Let me know if you've tested this a bunch and are confident it works fine for you, so you consider it ready for merge.

File: library/external/textdrawable/build.gradle
Comment (Turubaev) [reply library/external/textdrawable/build.gradle:3]: I was unable to build properly and launch an app, with this duplicated line and after removing it, building was success. I am not sure how it affects and can't remember exact error, unfortunately. Do you think we should return back this duplicated line ?

**Comment by Turubaev** [comment]
> Thanks! With these changes, it no longer notified for jitsi calls for me. Looks like I got `JitsiEventContent(type=jitsi, name=Jitsi)`, i.e. with uppercase name rather then what you're checking. I was able to fix by adding `lowercase()` calls to the check, so it matches both lowercase and uppercase scenarios:
> 
> ```
> content?.toModel<JitsiEventContent>()?.type?.lowercase() == IS_JITSI_CALL && content.toModel<JitsiEventContent>()?.name?.lowercase() == IS_JITSI_CALL
> ```
> 
> Otherwise seems to work fine now. I think I'm fine with merging, but I'll probably want to do some code refactoring right after to reduce possible future merge conflicts (I've learnt some tricks which probably doesn't make sense to expect from new contributors to apply).
> 
> Let me know if you've tested this a bunch and are confident it works fine for you, so you consider it ready for merge.

Yeap, I think modifying to lowercase would be much better, to ensure that strings compared correctly. Tested after applying this changes, everything looks good. From my perspective, it is ready to be merged

**Comment by alexander-potemkin** [comment]
> Otherwise seems to work fine now. I think I'm fine with merging

@SpiritCroc thank you very much! I don't mean to push or cause any other kind of inconvenience - I understand you are doing and will be doing it when you can. Just to plan things - do you have any ideas when you believe you can have this merged?

 > I'll probably want to do some code refactoring right after to reduce possible future merge conflicts (I've learnt some tricks which probably doesn't make sense to expect from new contributors to apply).

And I can't help myself and ask: what is the idea in general?
My experience is that keeping a fork up to date is quite a task, so in my personal experience I try keeping abstractions solid and keep my code aside as much as possible, keeping in mind that fork and merge would be a `diff` / `patch` work after all, but I wonder what is your experience and tools set in that case? If you don't mind me asking?

**Comment by SpiritCroc** [comment]
> @SpiritCroc thank you very much! I don't mean to push or cause any other kind of inconvenience - I understand you are doing and will be doing it when you can. Just to plan things - do you have any ideas when you believe you can have this merged?

No ETAs.

> And I can't help myself and ask: what is the idea in general? My experience is that keeping a fork up to date is quite a task, so in my personal experience I try keeping abstractions solid and keep my code aside as much as possible, keeping in mind that fork and merge would be a `diff` / `patch` work after all, but I wonder what is your experience and tools set in that case? If you don't mind me asking?

There's a bunch of tricks, I haven't really started writing them down yet. Some things

- Use your own files instead of upstream files as much as possible (kotlin extension functions are great for this!), unless you want to be sure you don't miss any updates to the upstream login in that file. Exaggerated example to keep upstream impact low: https://github.com/SchildiChat/schildichat-android-next/commit/53667548a47fb544d92c0af86e91431e2193cf8d
- When wrapping upstream code into additional blocks (if-else or whatever), don't mind code-style, just keep the upstream indention as is.
- For some things, may be useful to write scripts to revert some changes to upstream before the merge, and restore your own changes after. See `pre_merge.sh` and `post_merge.sh` in this repo.

**Comment by alexander-potemkin** [comment]
> There's a bunch of tricks, I haven't really started writing them down yet. Some things
> 
> * Use your own files instead of upstream files as much as possible (kotlin extension functions are great for this!), unless you want to be sure you don't miss any updates to the upstream login in that file. Exaggerated example to keep upstream impact low: [SchildiChat/schildichat-android-next@5366754](https://github.com/SchildiChat/schildichat-android-next/commit/53667548a47fb544d92c0af86e91431e2193cf8d)
> * When wrapping upstream code into additional blocks (if-else or whatever), don't mind code-style, just keep the upstream indention as is.
> * For some things, may be useful to write scripts to revert some changes to upstream before the merge, and restore your own changes after. See `pre_merge.sh` and `post_merge.sh` in this repo.

Thank you! I had something similar from my experience, but couldn't help myself with code-style things... That would help me to stay in line :)

> No ETAs.

Understood, makes perfect sense.

If there is something we could do - please, let us know! We would really need that working the sooner - the better, so if there is anything - anything at all - that could be done from my side - just ping me.

**Comment by SpiritCroc** [comment]
Out of curiosity, what makes this feature urgent in your eyes? I'm a bit surprised that people invest time on jitsi at all with Element call on the horizon. (Not that that would affect how much time my weekend gives me, just curious)

**Comment by alexander-potemkin** [comment]
> Out of curiosity, what makes this feature urgent in your eyes? I'm a bit surprised that people invest time on jitsi at all with Element call on the horizon. (Not that that would affect how much time my weekend gives me, just curious)

Sure, to keep things short: Element X is not exactly there yet, and I guess it will take them like half-of-the-year to stabilize and/or reach features parity. And missing the calls is something very  much annoying, especially if those group calls are actually a 1:1 calls - that breaks ability to work normally.

The thing is that 'groups' are happening at Matrix when you add a third something to 1:1 chat. That something doesn't have to be a person - it could be a bot doing some function you need for work. So, immediately, after you try to enhance your chat with a tool, you lose the ability to make calls.
And it goes without saying, that having a group / conference calls feels just natural.

So, at the end of the day we end up with a custom development. Android team seems like ignoring anything related to that functionality ([issue at meta](https://github.com/element-hq/element-meta/issues/2385#issuecomment-2395102440), [issue at android](https://github.com/element-hq/element-android/issues/7761)).
I understand the Element/Matrix team is now passing over complicated times, merging the code from someone outside still requires some time and resources, but that particular issue is like a blank spot, or at least that's how it feels to me.

I always liked SchildiChat approach (you might remember - I was asking for an iOS port earlier) so I decided this time we shall try our luck with SchildiChat and I'm glad to see that you are responding.

I understand that the amount of time in days is limited and that to re-organize code takes time, but it's still our best option, and I'm absolutely grateful for everything you do!

**Comment by SpiritCroc** [comment]
I'm surprised this was closed, is it no longer needed to merge?

FWIW, I had some looks at merging it before but I had to address some concerns before, and with this PR being rather big and this feature not being a priority for the project I never got to finish. My concerns were 1) merge conflicts, which we *may* get around if upstream doesn't update the codebase much anymore but we don't know, maybe possible to defer until later, and 2) need to be dead sure that the feature doesn't regress anything, which would be easiest if *all* functionality introduced by this patch was limited to having the setting enabled, which however is *not* the case. So to be 100% happy with this change I'd still have to basically go through every file which is almost the same effort as writing the feature myself in the first place, which is a bit much to ask from me for this codebase in maintenance mode.

My current TODO looked as following from how far I got:
- [ ] Revise NotificationUtils
    - Really needed as entry point?
    - Can we move to separate files?
- [ ] MainActivity jitsiCallIntent - why is it not just a normal call
- [ ] cancelNotificationMessage on CALL_ACTION_REJECT should maybe be separate
- [ ] ...

So while I'm happy to save my precious time and stop looking into this, since the PR is now closed, I didn't expect it since the feature seemed to be important to the PR originators. So I figured, I'd share the state it ended up here, so if anyone feels the need to pick this up they can give it better chances to succeed:

- I pushed what I already did for cleanup here: https://github.com/SchildiChat/SchildiChat-android/commit/c2a6aae597f1d7fdaa0106fb12fe7c092e2e1b4d - if you want to get this PR going again this may be a good starting point
- To make review on my side more straightforward, some refactoring would be appreciated
  - New functionality in new files
  - If you have to change existing functionality, safe-guard the changed functionality by the feature flag
  - Don't follow too much what upstream does (compare above cleanup commit removing some things which we do not care about downstream, since that'll just introduce potential merge conflict areas for now benefit)

**Comment by Turubaev** [comment]
This PR doesn't work as expected, therefore there is no need in merge

**Comment by alexander-potemkin** [comment]
Thank you, @SpiritCroc , appreciate all your time and efforts! Special thanks for sharing the actual state and concerns!

Would it be correct to assume, that once the concerns would be addressed: isolate the whole feature with the feature flag - first, and isolate changes in a separate files, trying to minimize upstream changes - second - then the PR is likely to be merged?

And, yeah, there was one thing missing from functionality as well, as mentioned above - the app did not pick up default sound notification, making calls actually silent.

If that is correct, I would rather like to try again to make this feature merged. If not - understood, no problems, I understand you are taking care of the project on your free will. I hope, however, that I can make something to add it to the project, while keeping project maintainer - you - happy.

**Comment by SpiritCroc** [comment]
> Would it be correct to assume, that once the concerns would be addressed: isolate the whole feature with the feature flag - first, and isolate changes in a separate files, trying to minimize upstream changes - second - then the PR is likely to be merged?

I don't want to state any absolute "likeliness". Addressing these concerns will of course make it more likely to get it merged but I can't promise anything to the eventual outcome, since I can't predict what time and priorities allow on my side.

> If that is correct, I would rather like to try again to make this feature merged. If not - understood, no problems, I understand you are taking care of the project on your free will. I hope, however, that I can make something to add it to the project, while keeping project maintainer - you - happy.

Thank you :)

**Pull request closed without merging**
